chore(docs): update release_process docs (#6353)
mozilla/blurts-server: docs/release_process.md

The standard release interval for Monitor is one week, meaning there should be at least one new version of the Monitor web app on the [Production][prod] environment. However, since we've started doing 1-click deploy to production, our release cycle can become more and more frequent.

## Release to Stage

> Every commit to `main` is automatically deployed to the [Stage][stage] server via Github Actions. The image is built and uploaded to Github Artifact Registry. ArgoCD scans the registry and detects whenever a new image uploaded with a tag that is a short commit SHA. When one is detected, it triggers a new deployment.

Reviewer: Heh, and here I was thinking GAR stood for Google Artifact Repository :)

Do you happen to have a link to where Argo is configured to look for short commit SHAs in the repository, and if so, could you add it here? That would help for me to get a feel for how it all fits together.

- Build
- Unit Tests
- E2E Tests

A PR also needs at least one approval from an ENGR team member to be merged into `main`.

Once a PR is successfully merged:

1. Ensure that the merge commit in `main` branch passes all checks and a docker image is successfully deployed to GAR.
2. A webhook will send status messages into the `#fx-monitor-engineering` channel when the deployment is completed.

## Release to Production

### Daily pre-releases

> Github pre-releases are generated daily via a [daily-pre-release](daily-pre-release) GHA workflow.

Author: I didn't realize that was the expected workflow. I had assumed we were mostly promoting the pre-releases verified by QA unless there were later commits that needed to be released. @Vinnl thoughts here?

> Github pre-releases are generated daily via a [daily-pre-release](daily-pre-release) GHA workflow.

Reviewer: Yeah my recent prod deploys were relatively uneventful, so I've been using the pre-releases without needing to merge last-minute PRs.

(That said, I'm not sure what the value of the daily prereleases is. It's not _that_ much effort to create a new tag at release time, and I don't know what we'd ever use the intermediate releases for.)

### Update Production Environment Variables

> Environment variables are configured in 2 places:

Reviewer: ```suggestion
Environment variables are configured in 3 places:
```

- Secret env vars via [Secret Manager](https://console.cloud.google.com/security/secret-manager)

> - Non-secret env vars via Helm [ConfigMap](https://github.com/mozilla/webservices-infra/blob/e8301f4b257550787098a25b078da99765e7a81a/monitor/k8s/monitor-www)

Reviewer: In principle I think it's good to use permalinks usually, but in this case I think there might be a risk of someone trying to find the current value of a variable, and ending up with an old one.

And if the path does change in the future, I think we'd rather know right away by running into a dead link, than being under the impression that this documentation is still accurate.

```suggestion
- Non-secret env vars via Helm [ConfigMap](https://github.com/mozilla/webservices-infra/tree/main/monitor/k8s/monitor-www)
```

- Terraform-generated secrets

Any changes directly to production should be done as a pairing session with another engineer. By default, engineers do not have access to production-level secrets. In order to make changes:

- Share your screen with your pairing partner
- Use [PAM](https://console.cloud.google.com/iam-admin/pam/entitlements/) to request 'secret-add-access' access grant for the minimum length of time required to complete your update
- Copy the latest secrets version into a new version
- Update the necessary variables
- Review with your pairing partner before accepting

> New secrets are discovered automatically by ArgoCD every 5 minutes. You can also use annotations to force the discovery. However, new secrets will _not_ be automatically pulled into currently running pods. They will only be updated when the pods are restarted or the service redeployed (this includes new cron job instances).

Author: I'm going to remove, this is something I watched Brett (SRE) do but we may not have this level of permissions anyway

> To restart the pods for the monitor web application, navigate to [ArgoCD](https://webservices.argocd.global.mozgcp.net/applications?search=monitor) and select the project for the appropriate environment. Find the rollout resources `monitor-api-deployment` and `monitor-www`. On each node, click the hamburger menu and select "restart pods". ArgoCD will spin up new pods and terminate the existing ones. The new pods will have the updated environment variables.

Reviewer: > `monitor-api-deployment`

Huh, I wasn't aware of that rollout. What does that contain?

   - Input the tag created earlier (today's date, e.g., `2024.09.01`)
   - Click on `Run workflow` when ready
7. A webhook will send status messages into the `#fx-monitor-engineering` channel.
8. After successful deploy, conduct some basic sanity check:
   - Check sentry prod project for a spike in any new issues
   - Check [grafana dashboard][grafana-dashboard] for any unexpected spike in ops
   - Spot-check the site for basic functionality

Note the following caveats:

- All production releases must follow the date tag naming scheme or else they will not be deployed. The regex for image deployments is in the [monitor tenant definition file](https://github.com/mozilla/global-platform-admin/blob/main/tenants/monitor.yaml).
  - Example: `2024.09.01`; also can include additional numeric suffix preceded by a period, e.g. `2024.09.01.1` (for manually created deployments, typically hotfixes)

> - All dev releases must use the short commit SHA as the tag or else they will not be deployed

Reviewer: Ah, I get it now - thanks for the clarification!

- Only commits which have already been deployed to staging can be released to development or production environments. In practice this means that you cannot and should not attempt to deploy from a tag created on a release branch
<!-- 

> Note: The reason that this is the case currently is due to a race condition with the on-push triggered jobs release_retag_v2 and docker_build_deploy_v2. The release_retag_v2 action attempts to pull an image uploaded by docker_build_deploy_v2 immediately, even though the job is not complete. However, the new tag _will_ create a staging release (tagged by the short commit SHA) that will override the current staging environment. This could result in confusion about the state of staging environment and commit verification by QA.

Reviewer: Ah thanks for this, I was just about to ask this above :)

> A prod release could be forced by re-running the failed release_retag_v2 job after docker_build_deploy_v2 completes, but this is not recommended. If the changes in this release branch are not merged into main and included in subsequent releases, then they could be unintentionally reverted.

Reviewer: ```suggestion
A prod release could be forced by re-running the failed release_retag_v2 job after docker_build_deploy_v2 completes, but this should be the exception, and you should make sure to merge the release branch back into `main` afterwards to prevent it from being unintentionally reverted.

Also note that any migrations already in `main` should also be included in your release branch.
```

(We've had a number of cases where this would definitely have been safer than just deploying the latest `main`, or than reverting and re-applying other changes.)
